Strata/Transform/CallElim.lean

```diff
@@ -13,234 +13,322 @@ import Strata.Languages.Boogie.Verifier
+                          proc.spec.preconditions)
+                        (arg_subst ++ ret_subst)
+        -- generate assumes based on post-conditions, substituting procedure arguments and returns
+        let assumes := createAssumes postconditions
+                        (arg_subst ++ ret_subst)
+
+        return argInit ++ outInit ++ oldInit ++ asserts ++ havocs ++ assumes
       | _ => return [ st ]
 
-def callElimStmts (ss: List Boogie.Statement) (prog : Boogie.Program) (wf : WF.WFStatementsProp prog ss)
-  : CallElimM (List Boogie.Statement) := do match ss with
+def callElimStmts (ss: List Statement) (prog : Program)
+  -- (wf : WF.WFStmtsProp Expression prog ss)
+  : CallElimM (List Statement) := do match ss with
     | [] => return []
     | s :: ss =>
-      let ⟨wfs, wfss⟩ : (WF.WFStatementProp prog _) ×' _ := ListP.split wf
-      let s' := (callElimStmt s prog wfs)
-      let ss' := (callElimStmts ss prog wfss)
+      -- let ⟨wfs, wfss⟩ : (WF.WFStmtProp _ prog _) ×' _ := ListP.split wf
+      let s' := (callElimStmt s prog)
+      let ss' := (callElimStmts ss prog)
       return (← s') ++ (← ss')
 
 -- TODO: integrate the program parameter into the state as well
 -- Walk through dcls, eliminating procedure call in them
-def callElimL (dcls : List Boogie.Decl) (prog : Boogie.Program) (wf : WF.WFDeclsProp prog dcls)
-  : CallElimM (List Boogie.Decl) :=
+def callElimL (dcls : List Decl) (prog : Program)
+  : CallElimM (List Decl) :=
   match dcls with
   | [] => return []
   | d :: ds =>
-    let ⟨wfd, wfds⟩ : (WF.WFDeclProp prog _) ×' _ := ListP.split wf
     match d with
     | .proc p =>
-      return Boogie.Decl.proc { p with body := ← (callElimStmts p.body prog wfd.wfstmts) } :: (← (callElimL ds prog wfds))
-    | _       => return d :: (← (callElimL ds prog wfds))
+      return Decl.proc { p with body := ← (callElimStmts p.body prog ) } :: (← (callElimL ds prog))
+    | _       => return d :: (← (callElimL ds prog))
 
 -- walk through all procedure bodies
-def callElim' (p : WF.WFProgram) : CallElimM Boogie.Program := return { decls := (← (callElimL p.self.decls p.self p.prop.wfdecls)) }
-
-def callElim (p : WF.WFProgram)
-  : Boogie.Program :=
-  (StateT.run (callElim' p) { }).fst
+def callElim' (p : Program) : CallElimM Program := return { decls := (← (callElimL p.decls p)) }
+
+@[simp]
+def runCallElimWith' {α : Type} (p : α) (f : α → CallElimM β) (s : BoogieGenState):
+  Except Err β × BoogieGenState :=
+  (StateT.run (f p) s)
+
+@[simp]
+def runCallElimWith {α : Type} (p : α) (f : α → CallElimM β) (s : BoogieGenState):
+  Except Err β :=
+  (runCallElimWith' p f s).fst
+
+-- run call elimination with an empty counter state (e.g. starting from 0)
+@[simp]
+def runCallElim {α : Type} (p : α) (f : α → CallElimM β):
+  Except Err β :=
+  runCallElimWith p f .emp
+
+def callElim (p : Program)
+  : Program :=
+  match (runCallElim p callElim') with
+  | .ok res => res
+  | .error e => panic! e
```

Comment:
I'd prefer to avoid `panic!` and say `callElim p := runCallElim p callElim'`.

Reply:
I believe using `callElim p := runCallElim p callElim'` does not unwrap the result from the `Except` monad. The match statement here is for extracting the actual `Program` that is transformed, and the `panic!` seems necessary for handling the error case. It is only used in `Examples.lean`, where it will give a `Program` upon success and raise an error when the transformation fails.

Reply:
Yeah, that was my intent: to require the caller to do a match on the result. I think it might be reasonable, in `Examples.lean`, to have a wrapper which calls `panic!`, since you know it won't actually occur for that specific example.

Reply:
I see. In fact this `callElim` function is only used in `Examples.lean`. What do you think about moving this function to that file (as it is now)? Since there are multiple examples, it is cleaner to have a single function with match instead of doing an explicit match for every example, and that was the intent for `callElim`.